ecdsa.go

```diff
@@ -0,0 +1,157 @@
+package cose
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+
+	"golang.org/x/crypto/cryptobyte"
+	"golang.org/x/crypto/cryptobyte/asn1"
+)
+
+// I2OSP - Integer-to-Octet-String primitive converts a nonnegative integer to
+// an octet string of a specified length `len(buf)`, and stores it in `buf`.
+// I2OSP is used for encoding ECDSA signature (r, s) into byte strings.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8017#section-4.1
+func I2OSP(x *big.Int, buf []byte) error {
+	if x.Sign() < 0 {
+		return errors.New("I2OSP: negative integer")
+	}
+	if len(x.Bits()) > len(buf) {
+		return errors.New("I2OSP: integer too large")
+	}
+	x.FillBytes(buf)
+	return nil
+}
+
+// OS2IP - Octet-String-to-Integer primitive converts an octet string to a
+// nonnegative integer.
+// OS2IP is used for decoding ECDSA signature (r, s) from byte strings.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8017#section-4.2
+func OS2IP(x []byte) *big.Int {
+	return new(big.Int).SetBytes(x)
+}
+
+// ecdsaKeySigner is a ECDSA based signer with golang built-in keys.
+type ecdsaKeySigner struct {
+	alg Algorithm
+	key *ecdsa.PrivateKey
+}
+
+// Algorithm returns the signing algorithm associated with the private key.
+func (es *ecdsaKeySigner) Algorithm() Algorithm {
+	return es.alg
+}
+
+// Sign signs digest with the private key, possibly using entropy from rand.
+// The resulting signature should follow RFC 8152 section 8.1.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-8.1
+func (es *ecdsaKeySigner) Sign(rand io.Reader, digest []byte) ([]byte, error) {
+	r, s, err := ecdsa.Sign(rand, es.key, digest)
+	if err != nil {
+		return nil, err
+	}
+	return encodeECDSASignature(es.key.Curve, r, s)
+}
+
+// ecdsaKeySigner is a ECDSA based signer with a generic crypto.Signer.
+type ecdsaCryptoSigner struct {
+	alg    Algorithm
+	key    *ecdsa.PublicKey
+	signer crypto.Signer
+}
+
+// Algorithm returns the signing algorithm associated with the private key.
+func (es *ecdsaCryptoSigner) Algorithm() Algorithm {
+	return es.alg
+}
+
+// Sign signs digest with the private key, possibly using entropy from rand.
+// The resulting signature should follow RFC 8152 section 8.1.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-8.1
+func (es *ecdsaCryptoSigner) Sign(rand io.Reader, digest []byte) ([]byte, error) {
+	sig, err := es.signer.Sign(rand, digest, nil)
+	if err != nil {
+		return nil, err
+	}
+	r, s, err := decodeECDSASignatureASN1(sig)
+	if err != nil {
+		return nil, err
+	}
+	return encodeECDSASignature(es.key.Curve, r, s)
+}
+
+// decodeECDSASignatureASN1 decodes (r, s) from ASN.1 encoded signature.
+//
+// Code copied from https://github.com/golang/go/blob/go1.18/src/crypto/ecdsa/ecdsa.go#L338-L354
+func decodeECDSASignatureASN1(sig []byte) (r, s *big.Int, err error) {
+	r, s = &big.Int{}, &big.Int{}
+	var inner cryptobyte.String
```

Comment:
We depend on the `x/crypto` module just for reading these ASN.1 encoded big integers. I don't have a strong opinion here, but I would probably just use the build-in `encoding/asn1` package and remove the `x/crypto` dependency:

```go
func decodeECDSASignatureASN1(sig []byte) (r, s *big.Int, err error) {
    var esig  struct {
        R, S *big.Int
    }
    if _, err := asn1.Unmarshal(sig, &esig); err != nil {
        return nil, nil, errors.New("ecdsa: invalid signature: invalid ASN.1 encoding")
    }
    return esig.R, esig.S, nil
}
```

Go crypto package does use it, but I believe the main driver is to avoid making `crypto` depend on `encoding\asn1` for the sake of having a sane dependency tree.

Reply:
The following code shows that two implementations are not equivalent.

```go
package main

import (
	easn1 "encoding/asn1"
	"errors"
	"fmt"
	"math/big"

	"golang.org/x/crypto/cryptobyte"
	"golang.org/x/crypto/cryptobyte/asn1"
)

func main() {
	sig := encode()

	fmt.Println(decodeUsingCryptoByte(sig))
	fmt.Println(decodeUsingASN1Unmarshal(sig))
}

func encode() []byte {
	r := new(big.Int).SetInt64(42)
	s := new(big.Int).SetInt64(0)

	var b cryptobyte.Builder
	b.AddASN1(asn1.SEQUENCE, func(b *cryptobyte.Builder) {
		b.AddASN1BigInt(r)
		b.AddASN1BigInt(s)
		b.AddASN1OctetString([]byte("malicious")) // add some malicious content here
	})
	res, err := b.Bytes()
	if err != nil {
		panic(err)
	}
	return res
}

func decodeUsingCryptoByte(sig []byte) (r, s *big.Int, err error) {
	r, s = &big.Int{}, &big.Int{}
	var inner cryptobyte.String
	input := cryptobyte.String(sig)
	if !input.ReadASN1(&inner, asn1.SEQUENCE) ||
		!input.Empty() ||
		!inner.ReadASN1Integer(r) ||
		!inner.ReadASN1Integer(s) ||
		!inner.Empty() {
		return nil, nil, errors.New("ecdsa: invalid signature: invalid ASN.1 encoding")
	}
	return
}

func decodeUsingASN1Unmarshal(sig []byte) (r, s *big.Int, err error) {
	var esig struct {
		R, S *big.Int
	}
	if _, err := easn1.Unmarshal(sig, &esig); err != nil {
		return nil, nil, errors.New("ecdsa: invalid signature: invalid ASN.1 encoding")
	}
	return esig.R, esig.S, nil
}
```

The above code outputs:

```
<nil> <nil> ecdsa: invalid signature: invalid ASN.1 encoding
42 0 <nil>
```

Reply:
However, since we are decoding from a trusted source just like the boringcrypto you have mentioned, I will re-implement this using the `asn1.Unmarhsal` method.